cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowDispatch.qll

```diff
@@ -231,10 +231,36 @@ private predicate functionSignature(Function f, string qualifiedName, int nparam
  * Holds if the set of viable implementations that can be called by `call`
  * might be improved by knowing the call context.
  */
-predicate mayBenefitFromCallContext(CallInstruction call, Function f) { none() }
+predicate mayBenefitFromCallContext(CallInstruction call, Function f) {
+  mayBenefitFromCallContext(call, f, _)
+}
+
+/**
+ * Holds if `call` is a call through a function pointer, and the pointer
+ * value is given as the `arg`'th argument to `f`.
+ *
+ * Note that `f` may be several layers up through the call chain.
```

Comment:
Is this comment wrong? The dataflow lib only supports one-level call contexts, so I'm not sure what this comment would mean.

Reply:
As long as `.flowsFrom` is restricted to local flow then the code looks fine.

Reply:
`.flowsFrom` does actually follow flow backward from callee to caller. Is that a problem?

Reply:
The dataflow lib assumes that `f` is the enclosing callable of `call`. I think any tuples for which that isn't the case will just end up causing spurious calculation - hopefully mostly harmless, but it might hurt performance a bit. I don't _think_ it'll do anything directly wrong.

Reply:
<!-- Error while uploading 2021-07-14-151654336.mp4 -->

Reply:
Summary of our Zoom call: The call context relation is very big, and we should really not include any unnecessary tuples. Since the dataflow library only has one level of call context, any `f` that is not the enclosing callable of `call` won't help resolve any calls.

It doesn't look like it's a problem currently. This may be because:
- We still have a lot of dataflow improvements to make
- C/C++ programmers don't use _that_ many function pointers (at least in the projects we've tested on currently).

 I'll create a PR that adds a conjunct that restricts `f` to be the enclosing call context of `call`.

Reply:
Fix is here: https://github.com/github/codeql/pull/6280